### PR TITLE
SAI attributes for path tracing for midpoint node .

### PR DESCRIPTION
**Pull request by aparnapattekar-gh**
Path tracing provides a record of the packet path as a sequence of interface ids. In addition, it provides a record of end-to-end delay, per-hop delay, and load on each egress interface along the packet delivery path.

Following draft describes path tracing in detail.
https://datatracker.ietf.org/doc/draft-filsfils-spring-path-tracing

This change adds SAI attributes to configure path tracing on mid point node.

### CODE REVIEW

**Comment by rlhui** [comment]
@aparnapattekar-gh - please join SAI meeting on 8/10 to present this  if available. Thanks.

**Comment by aparnapattekar-gh** [comment]
Hi Rita,What time is it? Can you/someone please forward the invite to me?Thanks,AparnaOn Aug 9, 2023, at 8:29 PM, Rita Hui ***@***.***> wrote:﻿
@aparnapattekar-gh - please join SAI meeting on 8/10 to present this  if available. Thanks.

**Comment by rlhui** [comment]
Hi,

It’s starting in 10min… Below is the meeting info.  Please let me know your email address?


SAI - OCP Project Call

Every Thursday 8am - 9:00am PST



OCP Project Call - SAI



SAI - OCP Project Call

Please join my meeting from your computer, tablet or smartphone.

https://global.gotomeeting.com/join/965362485<https://www.google.com/url?q=https%3A%2F%2Fglobal.gotomeeting.com%2Fjoin%2F965362485&sa=D&usd=2&usg=AFQjCNEDWa4Z5rrRMCbp2wR66Qj7wBtWOA<https://global.gotomeeting.com/join/965362485%3chttps:/www.google.com/url?q=https%3A%2F%2Fglobal.gotomeeting.com%2Fjoin%2F965362485&sa=D&usd=2&usg=AFQjCNEDWa4Z5rrRMCbp2wR66Qj7wBtWOA>>



You can also dial in using your phone.

United States: +1 (646) 749-3129

Access Code: 965-362-485



More phone numbers

Australia: +61 2 9087 3604

Austria: +43 7 2081 5427

Belgium: +32 28 93 7018

Denmark: +45 32 72 03 82

Finland: +358 942 72 1060

France: +33 170 950 594

Germany: +49 691 7489 928

Ireland: +353 16 572 651

Italy: +39 0 247 92 13 01

Netherlands: +31 207 941 377

New Zealand: +64 9 280 6302

Norway: +47 21 93 37 51

Spain: +34 932 75 2004

Sweden: +46 853 527 836

Switzerland: +41 225 4599 78

United Kingdom: +44 330 221 0088



First GoToMeeting? Let's do a quick system check:

https://link.gotomeeting.com/system-check<https://www.google.com/url?q=https%3A%2F%2Flink.gotomeeting.com%2Fsystem-check&sa=D&usd=2&usg=AFQjCNHaAWe88qKOvkD06kazCQdfCIgc9g<https://link.gotomeeting.com/system-check%3chttps:/www.google.com/url?q=https%3A%2F%2Flink.gotomeeting.com%2Fsystem-check&sa=D&usd=2&usg=AFQjCNHaAWe88qKOvkD06kazCQdfCIgc9g>>

**Comment by ahsalam** [comment]
ahabdels@cisco.com
apjoshi@cisco.com

**Comment by rlhui** [comment]
Thanks.   wondering if you are able to join this meeting today (happening in 2min)?

Rita

**Comment by ahsalam** [comment]
Yes. we will join and present the PR.

**Comment by ahsalam** [comment]
Hi @JaiOCP,

Many thanks for the interesting discussing and comments during the call. 

Regarding the comment of leveraging saitam.h for Path Tracing, that makes a lot of sense for some Path Tracing Nodes (e.g., Source Node) to create the probe instances of Path Tracing and configure the different parameters of each instance. All of probe instances can be maintained under the Path Tracing SAI object. 

However, the SAI attributes in this PR (Interface ID, and Time Template) are port attributes even if they are specific to Path Tracing. Adding the proposed port attributes under PT TAM Object, would require maintaining port list under the PT TAM to hold the two new port attributes. The list of ports under PT TAM will be just another a copy of the list of port under SAI Port. 
In addition, we would need a logic to keep these two same list of ports in sync: 
-	When port get assigned an ID, before we create the port under PT TAM, we have cross check the existence of the port in SAI port.
-	When a port is deleted from SAI port, we need to make sure it gets deleted from PT TAM as well. 
This makes the orchagent logic more complex and require lots of cross checks. 
Having the ID and Time template under SAI Port avoid these cross checks logic as they are property of the port. 

SAI TAM would for sure make sense for other roles like PT Source. 

Thanks 
Ahmed

**Comment by rlhui** [comment]
Can we please update the PR title to reflect that this PR is for midpoint function? thanks.

**Comment by aparnapattekar-gh** [comment]
> Can we please update the PR title to reflect that this PR is for midpoint function? thanks.

Done.

**Review by rck-innovium — APPROVED** [review]
The proposed PT Midpoint functionality looks good.

File: inc/saiport.h [hunk 1]
```diff
@@ -501,6 +501,25 @@ typedef enum _sai_port_host_tx_ready_status_t
...
+    /** Timestamp nanosecond bits [12:19] */
+    SAI_PORT_PATH_TRACING_TIMESTAMP_TYPE_12_19,
+
+    /** Timestamp nanosecond bits [16:23] */
+    SAI_PORT_PATH_TRACING_TIMESTAMP_TYPE_16_23,
+
+    /** Timestamp nanosecond bits [20:27] */
+    SAI_PORT_PATH_TRACING_TIMESTAMP_TYPE_20_27,
+
+} sai_port_path_tracing_timestamp_type_t;
```
Comment (kcudnik) [inline inc/saiport.h:521]: why those fields are in bits? can't this be a singel 32 i 64 bit integer?

Comment (ahsalam) [reply inc/saiport.h:521]: @kcudnik 
In Path Tracing [1], midpoint nodes record a truncated time stamp. The Truncated time stamp is 8-bits. 
This enum defines which 8-bits of the timestamp that the midpoint should record. 

[1]https://datatracker.ietf.org/doc/html/draft-filsfils-spring-path-tracing-04

**Comment by JaiOCP** [comment]
> Hi @JaiOCP,
> 
> Many thanks for the interesting discussing and comments during the call.
> 
> Regarding the comment of leveraging saitam.h for Path Tracing, that makes a lot of sense for some Path Tracing Nodes (e.g., Source Node) to create the probe instances of Path Tracing and configure the different parameters of each instance. All of probe instances can be maintained under the Path Tracing SAI object.
> 
> However, the SAI attributes in this PR (Interface ID, and Time Template) are port attributes even if they are specific to Path Tracing. Adding the proposed port attributes under PT TAM Object, would require maintaining port list under the PT TAM to hold the two new port attributes. The list of ports under PT TAM will be just another a copy of the list of port under SAI Port. In addition, we would need a logic to keep these two same list of ports in sync:
> 
> * When port get assigned an ID, before we create the port under PT TAM, we have cross check the existence of the port in SAI port.
> * When a port is deleted from SAI port, we need to make sure it gets deleted from PT TAM as well.
>   This makes the orchagent logic more complex and require lots of cross checks.
>   Having the ID and Time template under SAI Port avoid these cross checks logic as they are property of the port.
> 
> SAI TAM would for sure make sense for other roles like PT Source.
> 
> Thanks Ahmed

Here is a suggestion. Its only a 2 line change to use TAM.
1. Add new INT type SAI_TAM_INT_TYPE_PT in sai_tam_int_type_t
https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L479

2. Add SAI_PORT_ATTR_PATH_TRACING_TIMESTAMP_TYPE in  sai_tam_int_attr_t
https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L805

- TAM object is already part of port attribute. If TAM object is created for port then that port is enabled for path tracing.
https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saiport.h#L1592

- There is no need to create a new device id. There is already a device id attribute present and you can just use it.
https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L535

File: inc/saiport.h (hunk 1 above)
Comment (kcudnik) [reply inc/saiport.h:521]: so its a hardware limitation or protocol/design limitation ?

Comment (ahsalam) [reply inc/saiport.h:521]: It’s a protocol design (not a limitation). 
Every PT Midpoint along the packet delivery path records its Midpoint Compressed Data (MCD).
The MCD is designed to be very compact which includes interface ID and Truncated Timestamp (8bits)

The Regional collector (defined in the IETF draft) uses the midpoint truncated timestamp to construct the timestamp of the midpoint.

Comment (kcudnik) [reply inc/saiport.h:521]: lol seems confusing to me XD

**Comment by ahsalam** [comment]
> > Hi @JaiOCP,
> > Many thanks for the interesting discussing and comments during the call.
> > Regarding the comment of leveraging saitam.h for Path Tracing, that makes a lot of sense for some Path Tracing Nodes (e.g., Source Node) to create the probe instances of Path Tracing and configure the different parameters of each instance. All of probe instances can be maintained under the Path Tracing SAI object.
> > However, the SAI attributes in this PR (Interface ID, and Time Template) are port attributes even if they are specific to Path Tracing. Adding the proposed port attributes under PT TAM Object, would require maintaining port list under the PT TAM to hold the two new port attributes. The list of ports under PT TAM will be just another a copy of the list of port under SAI Port. In addition, we would need a logic to keep these two same list of ports in sync:
> > 
> > * When port get assigned an ID, before we create the port under PT TAM, we have cross check the existence of the port in SAI port.
> > * When a port is deleted from SAI port, we need to make sure it gets deleted from PT TAM as well.
> >   This makes the orchagent logic more complex and require lots of cross checks.
> >   Having the ID and Time template under SAI Port avoid these cross checks logic as they are property of the port.
> > 
> > SAI TAM would for sure make sense for other roles like PT Source.
> > Thanks Ahmed
> 
> Here is a suggestion. Its only a 2 line change to use TAM.
> 
> 1. Add new INT type SAI_TAM_INT_TYPE_PT in sai_tam_int_type_t
>    https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L479
> 2. Add SAI_PORT_ATTR_PATH_TRACING_TIMESTAMP_TYPE in  sai_tam_int_attr_t
>    https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L805
> 
> * TAM object is already part of port attribute. If TAM object is created for port then that port is enabled for path tracing.
>   https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saiport.h#L1592
> * There is no need to create a new device id. There is already a device id attribute present and you can just use it. But if you need port id then that can be added as port attribute
>   https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L535
> 
> Workflow will be something like this.
> 
> 1. Configure tam object with int type as PT
> 2. Configure tam attribute for choosing the PTP bits
> 3. Configure port id
> 4. Attach the tam object to port or set of ports: This enables the feature on that port.
> 5. Since this is a transit function no other tam attribute or object need to created

 @JaiOCP Many thanks for the proposal. It makes a lot of sense now. I haven't noticed that the TAM object is already part of port attribute. 
 
We will update the PR based on your proposal. But we would like to clarify few points of the proposal:

> * There is no need to create a new device id. There is already a device id attribute present and you can just use it. But if you need port id then that can be added as port attribute
>   https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L535

Do you mean that we can use the Device ID attribute to carry the  SAI_PORT_ATTR_PATH_TRACING_INTF ?

Or should we create a new to create attribute (SAI_PORT_ATTR_PATH_TRACING_INTF) in  sai_tam_int_attr_t as we did for SAI_PORT_ATTR_PATH_TRACING_TIMESTAMP_TYPE ?

**Comment by aparnapattekar-gh** [comment]
> > > > Hi @JaiOCP,
> > > > Many thanks for the interesting discussing and comments during the call.
> > > > Regarding the comment of leveraging saitam.h for Path Tracing, that makes a lot of sense for some Path Tracing Nodes (e.g., Source Node) to create the probe instances of Path Tracing and configure the different parameters of each instance. All of probe instances can be maintained under the Path Tracing SAI object.
> > > > However, the SAI attributes in this PR (Interface ID, and Time Template) are port attributes even if they are specific to Path Tracing. Adding the proposed port attributes under PT TAM Object, would require maintaining port list under the PT TAM to hold the two new port attributes. The list of ports under PT TAM will be just another a copy of the list of port under SAI Port. In addition, we would need a logic to keep these two same list of ports in sync:
> > > > 
> > > > * When port get assigned an ID, before we create the port under PT TAM, we have cross check the existence of the port in SAI port.
> > > > * When a port is deleted from SAI port, we need to make sure it gets deleted from PT TAM as well.
> > > >   This makes the orchagent logic more complex and require lots of cross checks.
> > > >   Having the ID and Time template under SAI Port avoid these cross checks logic as they are property of the port.
> > > > 
> > > > SAI TAM would for sure make sense for other roles like PT Source.
> > > > Thanks Ahmed
> > > 
> > > 
> > > Here is a suggestion. Its only a 2 line change to use TAM.
> > > 
> > > 1. Add new INT type SAI_TAM_INT_TYPE_PT in sai_tam_int_type_t
> > >    https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L479
> > > 2. Add SAI_PORT_ATTR_PATH_TRACING_TIMESTAMP_TYPE in  sai_tam_int_attr_t
> > >    https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L805
> > > 
> > > 
> > > * TAM object is already part of port attribute. If TAM object is created for port then that port is enabled for path tracing.
> > >   https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saiport.h#L1592
> > > * There is no need to create a new device id. There is already a device id attribute present and you can just use it. But if you need port id then that can be added as port attribute
> > >   https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L535
> > > 
> > > Workflow will be something like this.
> > > 
> > > 1. Configure tam object with int type as PT
> > > 2. Configure tam attribute for choosing the PTP bits
> > > 3. Configure port id
> > > 4. Attach the tam object to port or set of ports: This enables the feature on that port.
> > > 5. Since this is a transit function no other tam attribute or object need to created
> > 
> > 
> > @JaiOCP Many thanks for the proposal. It makes a lot of sense now. I haven't noticed that the TAM object is already part of port attribute.
> > We will update the PR based on your proposal. But we would like to clarify few points of the proposal:
> > > * There is no need to create a new device id. There is already a device id attribute present and you can just use it. But if you need port id then that can be added as port attribute
> > >   https://github.com/opencomputeproject/SAI/blob/39b7fe093d3f58f2b5bb037ac8dba94bf92f74ab/inc/saitam.h#L535
> > 
> > 
> > Do you mean that we can use the Device ID attribute to carry the SAI_PORT_ATTR_PATH_TRACING_INTF ?
> > Or should we create a new to create attribute (SAI_PORT_ATTR_PATH_TRACING_INTF) in sai_tam_int_attr_t as we did for SAI_PORT_ATTR_PATH_TRACING_TIMESTAMP_TYPE ?
> 
> There is a single attribute defined per TAM object for device id whereas if I understand PT requirement is to encode the egress interface meaning there is single TAM object container thats getting attached to multiple ports. I think for this usecase, adding the port attribute is the correct thing to do.

Hi @JaiOCP,

Thanks for the review and also for your comments/clarifications. Another question: If a TAM object attribute (e.g TIMESTAMP_TYPE) is modified, does the SAI implementation iterate over all the associated ports and modify it on the port? Also, handle other events, e.g if a port is removed, then does SAI remove the port from the list of ports that use this TAM object? In other words, does SAI maintain state to ensure the relationship between ports and TAM objects?

**Comment by ahsalam** [comment]
@rlhui @JaiOCP
Could we use some time at the beginning of next SAI meeting (August 17) to discuss the open questions related to this PR ?

**Comment by JaiOCP** [comment]
1. If time stamp attribute in TAM has changed then all the associated port objects need to walked by SAI adapter. This is no different than from e.g. nexthop walks for a given nexthop group or route object. There is no state maintained in SAI.2. If port is removed then It’s reference to TAM object is also removed. Again there is no state maintained in SAI adapter.More we can discuss in meeting.Regards,-JaiOn Aug 15, 2023, at 9:42 PM, Ahmed Abdelsalam ***@***.***> wrote:﻿
@rlhui @JaiOCP
Could we use some time at the beginning of next SAI meeting (August 17) to discuss the open questions related to this PR ?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

This electronic communication and the information and any files transmitted with it, or attached to it, are confidential and are intended solely for the use of the individual or entity to whom it is addressed and may contain information that is confidential, legally privileged, protected by privacy laws, or otherwise restricted from disclosure to anyone else. If you are not the intended recipient or the person responsible for delivering the e-mail to the intended recipient, you are hereby notified that any use, copying, distributing, dissemination, forwarding, printing, or copying of this e-mail is strictly prohibited. If you received this e-mail in error, please return the e-mail to the sender, delete it from your computer, and destroy any printed copy of it.

**Review by JaiOCP — APPROVED** [review]
Thanks for taking care of comments.
Changes looks good to me.

**Review by kcudnik — APPROVED** [review]

**Comment by ahsalam** [comment]
@kcudnik We addressed all the received comments. The PR should be ready for merge.

**Comment by kcudnik** [comment]
if this was already approved on Thursday SAI meeting then we can merge it

**Comment by ahsalam** [comment]
> if this was already approved on Thursday SAI meeting then we can merge it

The only comment was from Jai. We updated the PR and he approved the changes already.

**Comment by aparnapattekar-gh** [comment]
> > if this was already approved on Thursday SAI meeting then we can merge it
> 
> @kcudnik The only comment was from Jai. We updated the PR and he approved the changes already.



> if this was already approved on Thursday SAI meeting then we can merge it

@kcudnik Hi Kamil, all comments have been addressed already. Can you please merge the PR?

**Comment by aparnapattekar-gh** [comment]
> > > if this was already approved on Thursday SAI meeting then we can merge it
> > 
> > 
> > @kcudnik The only comment was from Jai. We updated the PR and he approved the changes already.
> 
> > if this was already approved on Thursday SAI meeting then we can merge it
> 
> @kcudnik Hi Kamil, all comments have been addressed already. Can you please merge the PR?

Thanks Kamil for merging the PR.

**Pull request merged**